File: src/AzureDataFactory.TestingFramework/Models/TestFramework.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

using Azure.ResourceManager.DataFactory;
using AzureDataFactory.TestingFramework.Exceptions;
using AzureDataFactory.TestingFramework.Models.Activities.Base;
using AzureDataFactory.TestingFramework.Models.Base;
using AzureDataFactory.TestingFramework.Models.Pipelines;
using AzureDataFactory.TestingFramework.Models.Repositories;

namespace AzureDataFactory.TestingFramework.Models;

public class TestFramework
{
    private readonly bool _shouldEvaluateChildPipelines;
    public DataFactoryRepository Repository { get; }

    public TestFramework(string? dataFactoryFolderPath = null, bool shouldEvaluateChildPipelines = false)
    {
        Repository = dataFactoryFolderPath != null ? DataFactoryRepositoryFactory.ParseFromFolder(dataFactoryFolderPath) : new DataFactoryRepository();
        _shouldEvaluateChildPipelines = shouldEvaluateChildPipelines;
    }

    /// <summary>
    /// Get an enumerator to yield the evaluation of the pipeline activities one by one using the provided parameters. The order of activity execution is simulated based on the dependencies. Any expression part of the activity is evaluated based on the state of the pipeline.
    /// </summary>
    /// <param name="pipeline">The pipeline to evaluate.</param>
    /// <param name="parameters">The global and regular parameters to be used for evaluating expressions.</param>
    /// <returns></returns>
    /// <exception cref="PipelineParameterNotProvidedException">Thrown if a required pipeline parameter is not required</exception>
    /// <exception cref="PipelineDuplicateParameterProvidedException">Thrown if a pipeline parameter is provided more than once</exception>
    public IEnumerable<PipelineActivity> Evaluate(Pipeline pipeline, List<IRunParameter> parameters)
    {
        pipeline.ValidateParameters(parameters);
        var state = new PipelineRunState(parameters, pipeline.Variables);
        return Evaluate(pipeline.Activities.ToList(), state);
    }

    public ActivityEnumerator EvaluateWithEnumerator(Pipeline pipeline, List<IRunParameter> parameters)
    {
        return new ActivityEnumerator(Evaluate(pipeline, parameters));
    }

    /// <summary>
    /// Evaluates all pipeline activities using the provided parameters. The order of activity execution is simulated based on the dependencies. Any expression part of the activity is evaluated based on the state of the pipeline.
    /// </summary>
    /// <param name="pipeline">The pipeline to evaluate.</param>
    /// <param name="parameters">The global and regular parameters to be used for evaluating expressions.</param>
    /// <returns></returns>
    /// <exception cref="PipelineParameterNotProvidedException">Thrown if a required pipeline parameter is not required</exception>
    /// <exception cref="PipelineDuplicateParameterProvidedException">Thrown if a pipeline parameter is provided more than once</exception>
    public List<PipelineActivity> EvaluateAll(Pipeline pipeline, List<IRunParameter> parameters)
    {
        return Evaluate(pipeline, parameters).ToList();
    }

    public IEnumerable<PipelineActivity> Evaluate(PipelineActivity activity, PipelineRunState state)
    {
        return Evaluate(new List<PipelineActivity> { activity }, state);
    }

    public IEnumerable<PipelineActivity> Evaluate(List<PipelineActivity> activities, PipelineRunState state)
    {
        while (state.ScopedPipelineActivityResults.Count != activities.Count)
        {
            var anyActivityEvaluated = false;
            foreach (var activity in activities
                         .Where(activity => !state.ScopedPipelineActivityResults.Contains(activity))
                         .Where(activity => activity.AreDependencyConditionMet(state)))
            {
                var evaluatedActivity = (PipelineActivity)activity.Evaluate(state);
                if (evaluatedActivity is not IIterationActivity || (evaluatedActivity is ExecutePipelineActivity && !_shouldEvaluateChildPipelines))
                    yield return evaluatedActivity;

                anyActivityEvaluated = true;
                state.AddActivityResult(activity);

                if (activity is IIterationActivity)
                {
                    if (activity is ExecutePipelineActivity executePipelineActivity && _shouldEvaluateChildPipelines)
                    {
                        var pipeline = Repository.GetPipelineByName(executePipelineActivity.Pipeline.ReferenceName);

                        // Evaluate the pipeline with its own scope
                        foreach (var childActivity in Evaluate(pipeline, executePipelineActivity.GetChildRunParameters(state)))
                            yield return childActivity;
                    }
                    else if (activity is ControlActivity controlActivity)
                    {
                        foreach (var childActivity in controlActivity.EvaluateControlActivityIterations(state, Evaluate))
                            yield return childActivity;
                    }
                }
            }

            if (!anyActivityEvaluated)
            {
                throw new ActivitiesEvaluatorInvalidDependencyException("Validate that there are no circular dependencies or whether activity results were not set correctly.");
            }
        }
    }
}

File: src/python/data_factory_testing_framework/models/test_framework.py
from typing import Any, List

from data_factory_testing_framework.generated.models import Activity, DependencyCondition, PipelineResource, \
    UntilActivity, ForEachActivity, IfConditionActivity, ExecutePipelineActivity, ControlActivity

from data_factory_testing_framework.models.repositories.data_factory_repository_factory import \
    DataFactoryRepositoryFactory
from data_factory_testing_framework.models.state.pipeline_run_state import PipelineRunState


class TestFramework:

    def __init__(self, data_factory_folder_path: str = None, should_evaluate_child_pipelines: bool = False):
        self.repository = data_factory_folder_path is not None and DataFactoryRepositoryFactory.parse_from_folder(data_factory_folder_path)
        self.should_evaluate_child_pipelines = should_evaluate_child_pipelines

    def evaluate_activity(self, activity: Activity, state: PipelineRunState) -> List[Activity]:
        return self.evaluate_activities([activity], state)

    def evaluate_pipeline(self, pipeline: PipelineResource, state: PipelineRunState) -> List[Activity]:
        return self.evaluate_activities(pipeline.activities, state)

    def evaluate_activities(self, activities: List[Activity], state: PipelineRunState) -> List[Activity]:
        while len(state.scoped_pipeline_activity_results) != len(activities):
            any_activity_evaluated = False
            for activity in filter(
                    lambda a: a not in state.scoped_pipeline_activity_results and a.are_dependency_condition_met(state),
                    activities):
                evaluated_activity = activity.evaluate(state)
                if not self._is_iteration_activity(evaluated_activity) or (isinstance(evaluated_activity, ExecutePipelineActivity) and not self.should_evaluate_child_pipelines):
                    yield evaluated_activity

                any_activity_evaluated = True
                state.add_activity_result(activity)

                if self._is_iteration_activity(activity):
                    if isinstance(activity, ExecutePipelineActivity) and self.should_evaluate_child_pipelines:
                        execute_pipeline_activity: ExecutePipelineActivity = activity
                        pipeline = self.repository.get_pipeline_by_name(execute_pipeline_activity.pipeline.reference_name)

                        # Evaluate the pipeline with its own scope
                        for childActivity in self.evaluate_pipeline(pipeline, activity.get_child_run_parameters(state)):
                            yield childActivity

                    if isinstance(activity, ControlActivity):
                        control_activity: ControlActivity = activity
                        for childActivity in control_activity.evaluate_control_activity_iterations(state, self.evaluate_activities):
                            yield childActivity

            if not any_activity_evaluated:
                raise Exception(
                    "Validate that there are no circular dependencies or whether activity results were not set correctly.")

    @staticmethod
    def _is_iteration_activity(activity: Activity) -> bool:
        return (isinstance(activity, UntilActivity) or
                isinstance(activity, ForEachActivity) or
                isinstance(activity, IfConditionActivity) or
                isinstance(activity, ExecutePipelineActivity))
